text/0011-gem-checksum-verification.md

```diff
@@ -0,0 +1,184 @@
+
+If a user wishes to ensure that the environment they are using contains only gems installed from verified sources that match the checksums in the Gemfile.lock, the command `bundle pristine` should be run. The `bundle pristine` command installs every gem fresh from the .gem source, but does not fetch gems from the server or read the source index. Therefore, if a users wants all gems to be verified upon install, `bundle pristine` should be run after checksums are written to the Gemfile.lock.
+
+Example:
+
+```
+$ bundle pristine
+Installing rake 13.0.6
+…
+Installing rspec 3.12.0
+42 gems are missing checksums. Source integrity could not be verified.
+Use `bundle lock` to add checksums to Gemfile.lock.
+$ bundle lock --update
+Writing lockfile to path/to/Gemfile.lock
+$ bundle pristine
+Installing rake 13.0.6 (verified)
+…
+Installing rspec 3.12.0 (verified)
+$
+```
+
+If a gem that is about to be installed does not match the checksum, an error is generated.
+
+Example:
+
+```
+$ bundle install
+Installing rake 13.0.6
+Bundler cannot continue installing rake-13.0.6.
+The checksum for the downloaded `rake-13.0.6.gem` does not match the known checksum for the gem. 
+
+Expected: sha256-814828c34f1315d7e7b7e8295184577cc4e969bad6156ac069d02d63f58d82e8 from: Gemfile.lock:21:1 CHECKSUMS rake (13.0.6)
+Downloaded: sha256-69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b6 from:
+path/to/rake-13.0.6.gem SHA256 hexdigest
+
+To resolve this issue:
+Delete the downloaded gem files referenced above and run `bundle install`
+
+If you are sure that the downloaded gem is correct, you can run…
+(RFC TODO: Currently undecided on what the command should be here to allow unverified gems.)
+`bundle install --skip-verification` OR
+`bundle config set unverified ignore` OR
+Remove the offending line from CHECKSUMS in Gemfile.lock
+```
+
+Certain checksums will always be unavailable because the source does not provide a checksum. When a checksum is not available, a line can be added to the Gemfile.lock CHECKSUMS section that explicitly indicates that no checksum is recorded for that gem version. Future installs of gems with intentionally ignored checksums should not error or warn when the gem is installed. The **(RFC TODO)** command will allow idicating that a gem version should ignore checksum verifications. Alternatively, the SHA256 digest of the gem at install time can be recorded allowing the user to verify future installs do not deviate from the recorded checksum.
+
+Platforms may pose a problem for Gemfile.lock checksums. If a user has not added one of the platforms where they bundle their app, then checksum verification could be skipped. In order to avoid this security vulnerability, the following should happen automatically.
+
+1. When a gem is installed and no checksum is available in the lockfile, a warning should be printed indicating that unverified gems were installed.
```

Comment:
should this be printed even if there's a checksum coming from the remote source? if so, that feels like it would get annoying

Reply:
Yes, there's a conflict between storing the checksums from the .gem on install and not installing unverified gems in production when the platform is missing. I think as you commented below, we want to be transparent about this unless the bundle is frozen.

I think we need to add this feature then: If new checksums are recorded during install while the bundle is frozen, then fail.

Reply:
Hmm, let me refine what I said there, because I don't think it captures the whole picture and I think there are conflicts.

Let's just talk about the frozen bundle situation.

Here's the bundler docs about frozen:

> --frozen
> Do not allow the Gemfile.lock to be updated after this install. Exits non-zero if there are going to be changes to the Gemfile.lock.
> This option is deprecated in favor of the frozen setting.

Given the lock:
```
CHECKSUMS
  foo (1.0.0)
  bar (1.0.0-darwin-23)
```

When does installing the gem fail during a frozen bundle?

1. Install foo. A checksum is computed at install, but not recorded because the bundle is frozen.
2. Install bar (1.0.0-x86_64-linux) which is entirely missing from the CHECKSUMS section.
3. Install baz which is entirely missing from the CHECKSUMS section.
4. All of the above because if you're running in frozen mode, the docs say it will fail if any update would produce a change to the lock.

This is a real challenge because this is going to happen to a lot of people:
1. Update bundler version
2. Run bundle install (no checksums are written unless the gems are not already installed)
3. Commit and push to production
4. Frozen install fails because new checksums are found.

This means we need to make exceptions for checksums being found from gem installs even when frozen. Or we need to say "if you're running in frozen mode, all checksums must be recorded that can be recorded". This means running `bundle lock` before you commit and push and might require that we inform people when we first write the CHECKSUMS section when it didn't exist before.

Reply:
Would the be fixed in we started recording checksums only for fresh lockfiles? That means people with existing lockfiles (and thus everyone using frozen) would not be affected initially. We could also add a `bundle lock --add-checksums` flag for people to opt-in to the feature on an existing lockfile.